Add Plantuml variant (re #166)

VladimirAlexiev opened the pull request:
Reopens #166: by using better technology (PlantUML) we can improve the diagram a lot more easily, and make new diagrams.
IMHO the [PlantUML code](https://github.com/w3c/data-shapes/pull/471/files) is more readable than the [textual diagram](https://github.com/w3c/data-shapes/blob/gh-pages/shacl12-core/index.html#L1147), 
and it produces real arrows (doesn't depend on the `Class-Diagram-Arrows.png` kluge)

<img width="654" height="633" alt="image" src="https://github.com/user-attachments/assets/d269f79c-3e02-43d5-a867-be53b76e62e0" />

- added link `NodeShape->PropertyShape`
- broke the long text to 2 lines
- shown `PropertyGroup`
- added `sh:order` to `PropertyShape` and `PropertyGroup`

These are minor fixes, but the point is that with PlantUML we can maintain more easily, add more diagrams (eg Node Expressions), and make more kinds of diagrams.

[See SVG diagram rendered at plantuml server](https://www.plantuml.com/plantuml/svg/hLPBRzGm4BxdLrZbXiX6m2r2g6AHk1IAAE9IUfYDPvFGs8x6atOZuBzZEzuysatL0tkQn_VDZ3CFxRbb8EvKarIrH57KL3He1mLLoxrGg7P8Dh6tjMw1G0asRKUDg9XRcsSPmVschBc2udPJ69NTloqo2GofhQ15cocmZ9GvhMXUlqaBGvZLMk9XKx7BcnGDM2lER9LVUSAP02k29AwlNmmnRcx4xqHudO5Av6sWECQaUamkIEvj7jGv_8kHwD6J5i2W-pdgSlSB2xQN-n4wcOx1ywzaMYHU4A-w9NEohL4txZT4a0W5rtV0A3rvfWRUmSfyPqo3e0UyGckX3A-9ueYpJBKkXQ6GhG5TNWNBGBHuXrHpxvb7EL3zTuzwydVU3V-TfewCcW-cdP8-R4rXj6M2MhCLsr4MOMQGaHpqbTWR4dW0rJQOHrfZBCg9i3MgDHer-ye6vyDF8Go59_MQuP1Irs0w05C_TzotwBaV-kzzA738OsWbXo5Nc4U0NYl3UIBcu6mh8EyQctrHMnxoQiCh1ES9-77CHTm15sRHYd5qfdOiB3v26mt_gosVQxqrdLwhNKMt8mSq79v4H_VOI0V04cbCTO6wv6gDVVHFoPx67_sdquNpi3flmoKP1knZdAOLi8pkwV9yGsiwAj3nVshXV-Cm2BDVOkA5ydkrEX_nK2qtJinsq_aNvqI3ma05zTp38D5ItN9jj2SwhOZgC_dpBMo1grJY7hg6Vq3JXLhc-dHtdqMzyu9TaWn99EyCGl3fJkr2Tn_lWGVwq8zxT7h3Ff7fsiM61OkFN0R1OOCmBLXGJnRaCWpx-UvFcauNT6PSxd4oGaJwVkaI-ThZeidbVPaiAQaBidn5Bauz93b7BVrVW7y0) 
- Try out the links: they work and are relative
  - There's `skinparam topurl https://raw.githack.com/w3c/data-shapes/master/shacl12-core/index.html` in the `.puml` file but we should try to leave it out and see whether the links use as base the HTML URL (good) or the SVG URL (bad)
- There's a Makefile that produces SVG using a local `plantuml` installation
- We want to use a local installation instead of the plantuml server because the build shouldn't depend on an external server
  - https://github.com/plantuml/plantuml/issues/2289 and because of this rendering problem
- https://github.com/plantuml/plantuml/issues/2290 There's was no blue underline on the class name (Update: added with tags `<color:blue>` and `<u>`). Note the whole node is clickable and goes to the class definition

If you agree, I can:
- replace the textual diagram in the HTML with `<img>` tag
- remove `Class-Diagram-Arrows.png`

Please check all links for accuracy!

VladimirAlexiev committed:
minor editorial change

VladimirAlexiev committed:
plantuml version of diagram

VladimirAlexiev committed:
add manual `<color>` and `<u>` tags to class names

VladimirAlexiev committed:
add manual `<color>` and `<u>` tags to class names

YoucTagh commented on shacl12-core/images/SHACL-UML.puml:
```diff
@@ -0,0 +1,47 @@
+  [[#constraints Constraint parameters]], for example:
+  [[#closedConstraintComponent sh:closed]] : xsd:boolean or [[#syntax-rule-closed-datatype sh:ByTypes]]
+  [[#orConstraintComponent sh:or]], [[#andConstraintComponent sh:and]], [[#xoneConstraintComponent sh:xone]] : rdf:List
+  [[#notConstraintComponent sh:not]] : [[#shapes sh:Shape]]
```
```suggestion
  [[#ClosedConstraintComponent sh:closed]] : xsd:boolean or [[#syntax-rule-closed-datatype sh:ByTypes]]
  [[#OrConstraintComponent sh:or]], [[#AndConstraintComponent sh:and]], [[#XoneConstraintComponent sh:xone]] : rdf:List
  [[#NotConstraintComponent sh:not]] : [[#shapes sh:Shape]]
```

YoucTagh commented on shacl12-core/images/SHACL-UML.puml:
```diff
@@ -0,0 +1,47 @@
+  [[#NodeConstraintComponent sh:node]] : [[#node-shapes sh:NodeShape]]
+  ....
+  [[#name sh:name]] : xsd:string or rdf:langString
+  [[#desription sh:description]] : xsd:string or rdf:langString
```
```suggestion
  [[#description sh:description]] : xsd:string or rdf:langString
```

YoucTagh commented on shacl12-core/images/SHACL-UML.puml:
```diff
@@ -0,0 +1,47 @@
+}
+
+class "sh:PropertyGroup" as PropertyGroup [[#group sh:group]] {
+  [[#group sh:order]] : number
```
```suggestion
  [[#order sh:order]] : xsd:decimal
```

VladimirAlexiev replied:
1. https://raw.githack.com/w3c/data-shapes/master/shacl12-core/index.html#order doesn't say it can be used on Group. https://raw.githack.com/w3c/data-shapes/master/shacl12-core/index.html#group says this. So I used `#group` on purpose
2. https://github.com/w3c/data-shapes/issues/479

VladimirAlexiev committed:
Update shacl12-core/images/SHACL-UML.puml

Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>

VladimirAlexiev committed:
Update shacl12-core/images/SHACL-UML.puml

Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>

VladimirAlexiev commented:
@YoucTagh Thanks for your close reading!!!!
Now do you agree with the next steps:
- replace the textual diagram in the HTML with `<img>` tag
- Check how the links behave when embedded in HTML
- remove `Class-Diagram-Arrows.png`

TallTed commented:
As I noted in #166 — 
> Assuming that the diagram we finalize on does not include all features, it should include some `...` or `etc.` or similar arrows/boxes/line-items — so it's clear that we're not showing everything in the diagram (which I think is entirely reasonable, even with automatic tooling, given the large number of features involved).

Also, note that Accessibility review will demand textual description of each diagram, usable with screen readers and other assistive technologies. So it's in our best interest to avoid overly complex diagrams, even if they're easily produced.

VladimirAlexiev committed:
- sh:order: xsd:integer or xsd:decimal
- sh:path : predicate or list or blank node and link to #property-paths
- add ellipsis as per TallTed
- rdf:List -> "list" because SHACL doesn't require the rdf:List type
- add sh:nodeKind, sh:ignoredProperties

VladimirAlexiev committed:
add <color:blue><u> to PropertyGroup

VladimirAlexiev commented:
@TallTed @YoucTagh I added a few more changes, can you please recheck?
- But don't yet merge, I'm trying to figure out how to enact the links
- The links at https://raw.githubusercontent.com/VladimirAlexiev/data-shapes/refs/heads/issue-166-plantuml/shacl12-core/images/SHACL-UML.svg are active, though lead to the SVG itself, since I commented out
`skinparam topurl https://raw.githack.com/w3c/data-shapes/master/shacl12-core/index.html`
- The links don't work from inside the HTML https://raw.githack.com/VladimirAlexiev/data-shapes/refs/heads/issue-166-plantuml/shacl12-core/index.html#constraints-section  :-(

VladimirAlexiev committed:
replace textual diagram with PlantUML SVG diagram

YoucTagh reviewed (approved):
Thank you, the content is good for me. I won't merge it; maybe we can figure out how to make the links work.

VladimirAlexiev committed:
replace <img> with script that embeds SVG content

VladimirAlexiev commented:
@YoucTagh  the links work now: https://raw.githack.com/VladimirAlexiev/data-shapes/refs/heads/issue-166-plantuml/shacl12-core/index.html#constraints-section.

@TallTed I can't yet add an accessibility description (`<svg desc=`) because https://github.com/plantuml/plantuml/issues/2292.
Hopefully PlantUML will add such feature soon!

TallTed reviewed (approved):
Much improved. There's always more, but this is a big step forward.

YoucTagh commented:
@VladimirAlexiev  Can we merge this PR?

VladimirAlexiev commented:
@YoucTagh yes, but the end of index.html should be like this (related to #486)
```js
	<script type="text/javascript">

    // Load and inline SVG diagrams for clickable links
    // TODO: add <svg desc> for accessibility
    function svgInline(svg) {
      fetch('images/'+svg)
        .then(response => response.text())
        .then(svgText => {
          const container = document.getElementById(svg);
          container.innerHTML = svgText;
        })};

    svgInline('SHACL-UML.svg');
    svgInline('ValidationReport-UML.svg');
		
	</script>
```

YoucTagh commented:
@VladimirAlexiev, ok, thank you.

Once #487 is merged, we will update index.html and then merge it.

VladimirAlexiev committed:
add core-components-list (3 fields), sh:targetWhere, sh:codeIdentifier. Move sh:path to top of PropertyShape

VladimirAlexiev commented:
@YoucTagh I've added a few of the new features: core-components-list (3 fields), sh:targetWhere, sh:codeIdentifier:
https://raw.githack.com/VladimirAlexiev/data-shapes/issue-166-plantuml/shacl12-core/index.html#constraints-section

(note: I put back the link to PropertyGroup and it shows ok on
https://github.com/VladimirAlexiev/data-shapes/blob/issue-166-plantuml/shacl12-core/images/SHACL-UML.svg
but for some reason doesn't show on the preview link: I suspect a caching issue)

I tried this variant with extra links to NodeShape but I think the diagram gets too busy, so I omitted them (they are listed as fields)
<img width="719" height="786" alt="image" src="https://github.com/user-attachments/assets/d358d435-528f-4270-a89e-76ea4aae9616" />

VladimirAlexiev committed:
restore link to PropertyGroup

afs commented:
Same concern as https://github.com/w3c/data-shapes/pull/483#issuecomment-3416396981
